target/scripts/check-for-changes.sh

```diff
@@ -6,47 +6,54 @@
 # shellcheck source=./helpers/index.sh
 source /usr/local/bin/helpers/index.sh
 
-function _log_date
+REGEX_NEVER_MATCH="(?\!)"
+
+function _changedetector_notify
 {
-  date +"%Y-%m-%d %H:%M:%S"
+  local LEVEL="${1}"
+  shift 1
+
+  _notify "${LEVEL}" "(CHANGEDETECTOR) $(date +"%Y-%m-%d %H:%M:%S") ${*}"
 }
 
-LOG_DATE=$(_log_date)
-_notify 'task' "${LOG_DATE} Start check-for-changes script."
+function _changedetector_exit_error
+{
+  _changedetector_notify "${@}"
+  _changedetector_notify 'Aborting'
+
+  exit 1
+}
 
-# ? --------------------------------------------- Checks
+_changedetector_notify 'debug' 'Starting changedetector'
 
-cd /tmp/docker-mailserver || exit 1
+if ! cd /tmp/docker-mailserver &>/dev/null
+then
+  _changedetector_exit_error "Could not change into '/tmp/docker-mailserver/' directory"
+fi
 
-# check postfix-accounts.cf exist else break
 if [[ ! -f postfix-accounts.cf ]]
 then
-  _notify 'inf' "${LOG_DATE} postfix-accounts.cf is missing! This should not run! Exit!"
-  exit 0
+  _changedetector_exit_error "'/tmp/docker-mailserver/postfix-accounts.cf' is missing"
 fi
 
-# verify checksum file exists; must be prepared by start-mailserver.sh
+# checksum file must have been prepared by start-mailserver.sh
 if [[ ! -f ${CHKSUM_FILE} ]]
 then
-  _notify 'err' "${LOG_DATE} ${CHKSUM_FILE} is missing! Start script failed? Exit!"
-  exit 0
+  _changedetector_exit_error "'/tmp/docker-mailserver/${CHKSUM_FILE}' is missing"
 fi
 
-# ? --------------------------------------------- Actual script begins
-
-# determine postmaster address, duplicated from start-mailserver.sh
-# this script previously didn't work when POSTMASTER_ADDRESS was empty
-_obtain_hostname_and_domainname
-
-PM_ADDRESS="${POSTMASTER_ADDRESS:=postmaster@${DOMAINNAME}}"
-_notify 'inf' "${LOG_DATE} Using postmaster address ${PM_ADDRESS}"
+# sleep some more during startup to avoid conflicting writes
+sleep 10
 
-REGEX_NEVER_MATCH="(?\!)"
+until grep -qE "export POSTMASTER_ADDRESS=.+" /root/.bashrc
+do
+  sleep 5
+done
 
-# Change detection delayed during startup to avoid conflicting writes
-sleep 10
+_get_dms_environment_variables
 
-_notify 'inf' "$(_log_date) check-for-changes is ready"
+_changedetector_notify 'trace' "Running with postmaster address '${POSTMASTER_ADDRESS}'"
+_changedetector_notify 'trace' 'Changedetector is ready'
```

Comment:
I read the [associated commit message](https://github.com/docker-mailserver/docker-mailserver/pull/2486/commits/cce33f80d545c23eaddb67a4afb2a2bd9596e0d9):

> The env variable `POSTMASTER_ADDRESS` was set in a way that is not very clear to the reader and indeed possibly wrong.
> The old way used the default while the new way uses the correct variable from `/root/.bashrc`.
> 
> Therefore, a new function `_get_dms_environment_variables` was introduced which handles exporting vaiables from `/root/.bashrc`. This could be used in the future as well.

In what case would the value be different? We're still presently in sync with `start-mailserver.sh`:

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/start-mailserver.sh#L10-L19

Is there a change that is made to `/root/.bashrc` elsewhere? In what case is it wrong?

You also state that it's not very clear with the way it's being set, is that something that needs to be improved at `start-mailserver.sh` then too? How is this change more clear? With the previous content at least we had a comment stating that `check-for-changes.sh` would not work if `POSTMASTER_ADDRESS` was empty hence the helper script call to ensure `DOMAINNAME` was assigned for the fallback if no `POSTMASTER_ADDRESS` is set in container ENV, just like `start-mailserver.sh`?.

Now we're sleeping 5 or more seconds to depend on extracting what should be the same value?

---

Then you've added a new helper method `_get_dms_environment_variables` that interacts with `/root/.bashrc`.. but from what I understand casper introduced `/etc/dms-settings` to source for this specific purpose, it is used in `target/bin/` scripts at present. Aren't you duplicating logic with this method?

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/startup/setup-stack.sh#L41-L57

That method is called early on in `start-mailserver.sh`:

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/start-mailserver.sh#L102

and later after all the setup, we start the `check-for-changes.sh` service:

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/start-mailserver.sh#L197

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/startup/daemons-stack.sh#L150-L154

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/start-mailserver.sh#L263-L276

So I'm not sure why we'd be waiting on such an ENV with `sleep` when by the time this script runs, the dependency should already be written. Neither location should be using remote storage/volume causing latency issues AFAIK, both would be using the internal container write layer?

---

If you're concerned about maintenance with potential for a desync, perhaps use a separate PR to introduce a common helper script. I assume `_get_dms_environment_variables` is not necessary, there is no mention of any additional ENV we're depending on from it.

If you rather source `/etc/dms-settings`, have a comment that explicitly states we're only doing so for `POSTMASTER_ADDRESS`, assuming that there would be no conflict with any other vars used in this script (I haven't checked).

Reply:
With

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/start-mailserver.sh#L19

a new variable `PM_ADDRESS` is created, but at the same time, the variable `POSTMASTER_ADDRESS` is assigned its default value before this value is again assigned to `PM_ADDRESS`. This is useful in `start-mailserver.sh`, but no here, because the call to

https://github.com/docker-mailserver/docker-mailserver/blob/321ae744fa873ca300af83983b4f916ac8dc9c35/target/scripts/check-for-changes.sh#L126

expects `POSTMASTER_ADDRESS` to be set correctly (otherwise `/etc/aliases` has errors in it).

But you're correct, using `/etc/dms-settings` would have been (actually is) easier, I just forgot. I will update the parts of the code.